NPE in DefaultClientFactory.newClient

Version : 1.11

```
Caused by: java.lang.NullPointerException
        at com.linecorp.armeria.client.DefaultClientFactory.newClient(DefaultClientFactory.java:174)
        at com.linecorp.armeria.client.ClientBuilder.build(ClientBuilder.java:112)
```

I'm seeing this NPE trying to connect to the scheme `gproto+h2c`. It works locally and fails only on a uber jar.
Definitely something is missing or mis-configured, but the error is just too cryptic...

### Investigation results

The actual factory is a `UnaryGrpcClientFactory` (added in #3773) and its condition to accept a `clientType` is `clientType == UnaryGrpcClient.class`. Why is the other factory (`GrpcClientFactory`) not present?

It's reference via a ServiceLoader config file, which is present in the uber jar;
Actually both service loader config files are present (duplicate entries).
Fixing the jar building process to merge the service properties into a single `com.linecorp.armeria.client.ClientFactoryProvider` jar entry fixed the concrete issue.

### Finally

- It would have saved me time if I knew that a factory is missing : making the error more explicit in #3868 
- My failing to build an uber jar is not related to how Armeria is built, so won't fix anything related, but keeping it here for reference

--- a/core/src/main/java/com/linecorp/armeria/client/DefaultClientFactory.java
+++ b/core/src/main/java/com/linecorp/armeria/client/DefaultClientFactory.java
@@ -162,16 +162,15 @@ final class DefaultClientFactory implements ClientFactory {
         validateParams(params);
         final Scheme scheme = params.scheme();
         final Class<?> clientType = params.clientType();
-        // `factory` must be non-null because we validated params.scheme() with validateParams().
-        ClientFactory factory = null;
-        for (ClientFactory f : clientFactories.get(scheme)) {
-            if (f.isClientTypeSupported(clientType)) {
-                factory = f;
-                break;
+        for (ClientFactory factory : clientFactories.get(scheme)) {
+            if (factory.isClientTypeSupported(clientType)) {
+                return factory.newClient(params);
             }
         }
-        assert factory != null;
-        return factory.newClient(params);
+        // Since we passed validation, there should have been at least 1 factory for this scheme,
+        // but for some reason none of these passed the filter.
+        throw new IllegalStateException(
+                "No ClientFactory for scheme: " + scheme + " matched clientType: " + clientType);
     }
 
     @Override

--- a/grpc/src/test/java/com/linecorp/armeria/client/grpc/protocol/UnaryGrpcClientTest.java
+++ b/grpc/src/test/java/com/linecorp/armeria/client/grpc/protocol/UnaryGrpcClientTest.java
@@ -84,7 +84,7 @@ class UnaryGrpcClientTest {
     @ParameterizedTest
     @ArgumentsSource(UnsupportedGrpcSerializationFormatArgumentsProvider.class)
     void unsupportedSerializationFormat(SerializationFormat serializationFormat) {
-        assertThrows(AssertionError.class,
+        assertThrows(IllegalStateException.class,
                      () -> Clients.newClient(getUri(serializationFormat), UnaryGrpcClient.class));
     }
 

